[new attempt] update the guide rates for all the wells at the same time.

GitPaean opened the pull request.

GitPaean commented:
jenkins build this opm-simulators=3406 please

GitPaean commented:
jenkins build this opm-simulators=3406 please

GitPaean commented:
The testing results is different from what I saw locally. There is something weird going on.

GitPaean commented:
jenkins build this opm-simulators=3406 please

GitPaean commented:
`8f8d97716e5b95678270f7559fa37c8cb4099103` shows how the zero valued guide rates impact the opm-tests results. It only causes problem for some situations.  It does not cause problem for the targeting case. 

And the jenkins testing results without and with the above commit are in the following links. 

Without the commit, (lot of failures)
https://ci.opm-project.org/job/opm-common-PR-builder/3855/

With the commit, (pass)
https://ci.opm-project.org/job/opm-common-PR-builder/3857/

bska commented on src/opm/parser/eclipse/EclipseState/Schedule/Group/GuideRate.cpp:
```diff
@@ -334,4 +322,28 @@ double GuideRate::get_grvalue_result(const GRValState& gr) const
+    }
+
+    const double update_delay = config.model().update_delay();
+    this->time_to_update = (sim_time >= (last_update_time + update_delay));
```
```C++
this->guide_rates_expired = !(sim_time < previous_update_time + update_delay);
```

GitPaean replied:
I have to say without using `!` is actually more straightforward to me. It is totally sensible if we use a different name, then the following can be better. 

```c++
this->guide_rates_valid = (sim_time < previous_update_time + update_delay);
```

bska reviewed (commented):
The current approach will fail unless we're guaranteed that we **always** call `compute` with the full set of wells or groups.  Otherwise there will be at least one element in `values` for which the current `sim_time` is in the past and which will trigger value updating upon calling `compute()` on the next time step.

If we're okay with that trade-off then this PR looks to meet the requirements.  Otherwise I still think a transactional model with explicit `begin/add/update/commit` stages and a single update time for the full set of guide rates will be a more future proof approach.

GitPaean commented:
@bska , thanks for your comment. I think we will do the trade-off manner to avoid huge refactoring including some contents in the opm-simulators.  I will dig in a little bit more regarding how to properly handle the zero valued guide rate.  

We can do a refactoring in another attempt.

GitPaean commented:
jenkins build this opm-simulators=3406 please

GitPaean commented:
> The current approach will fail unless we're guaranteed that we always call compute with the full set of wells or groups. Otherwise there will be at least one element in values for which the current sim_time is in the past and which will trigger value updating upon calling compute() on the next time step.

YES.   This is totally correct.

GitPaean commented:
Besides the zero rated guide rate, WGRUPCON might also cause problem when interacting with `GUIDERAT`, because all the guide rates are saved with a value and `sim_time`.  But as long as we are aware of the problem, we can tackle it. 

A full refactoring trying to distinguish/separate `WGRUPCON` assigned guide rates, closed/stopped/shut wells, and `GUIDERAT` handled wells/groups, injecting wells, will be very good, not sure what priority we should give.  

The current way is pretty difficult to keep track of everything and to have full control.

bska commented:
jenkins build this opm-simulators=3406 please

joakim-hove commented:
I have not managed to look into this yet - and I'm going offline for holiday; but my short answer is that I really think this should be implemented as: One function which calculates *all rates simultaneously* - getting all the required potentials in one (or a couple) containers. If there are physical constraints on how things should be done  it makes sense to let the API follow those constraints (in my opinion).

bska commented:
jenkins build this opm-simulators=3406 please

GitPaean commented:
> I have not managed to look into this yet - and I'm going offline for holiday; but my short answer is that I really think this should be implemented as: One function which calculates all rates simultaneously - getting all the required potentials in one (or a couple) containers. If there are physical constraints on how things should be done it makes sense to let the API follow those constraints (in my opinion).

I understand and agree with it. The current way is one way need less code change, hopefully, we achieve above principles in a theoretical way, and get the results straight (more PRs are required).  We should do the refactoring afterwards, in such a way without changing anything regarding result and it is a pure code/interface refactoring. Because there are so many things needs to take care of, I assume there will be some struggles to make all the tests happy.

GitPaean commented:
jenkins build this opm-simulators=3406 please

GitPaean commented:
I do not understand the jenkins compilation error.  Rebasing to see.

GitPaean commented:
jenkins build this opm-simulators=3406 please

bska commented:
> I do not understand the jenkins compilation error. Rebasing to see.

Please do a full rebase too, not just close/reopen.

GitPaean commented:
> Please do a full rebase too, not just close/reopen.

It is fully rebased now.

GitPaean commented:
It looks like the results from the base master branch changed from Friday, which make the results with this PR not match again. 

And locally, they failed with `assertion` failure and it looks like Jenkins do not test `assertion` failures.

GitPaean commented:
Actually, I do not know why the Jinkins fail now. Locally, except the `assertion` failure, this PR produces the same results with the master branch.

GitPaean commented:
downstream jenkins looks fine. Let us try again.

GitPaean commented:
jenkins build this opm-simulators=3406 please

GitPaean commented:
Besides, the assertion failure related is the commented out line below, 
``` c++
--- a/opm/simulators/wells/WellGroupHelpers.cpp
+++ b/opm/simulators/wells/WellGroupHelpers.cpp
@@ -945,7 +945,10 @@ namespace WellGroupHelpers
         const double my_guide_rate = guideRate(name, always_included_child);
         const Group& parent_group = schedule_.getGroup(parent(name), report_step_);
         const double total_guide_rate = guideRateSum(parent_group, always_included_child);
-        assert(total_guide_rate >= my_guide_rate);
+        // assert(total_guide_rate >= my_guide_rate);
+        if (total_guide_rate < my_guide_rate) {
+            std::cout << " name " << name << " total_guide_rate " << total_guide_rate << " my_guide_rate " << my_guide_rate << std::endl;
+        }
```

With the debugging output it outputs the following for the master branch and with this PR
```
 616  name OPL1 total_guide_rate 0 my_guide_rate 544.495
 617  name OPL2 total_guide_rate 0 my_guide_rate 5336.21
 618  name OPL3 total_guide_rate 0 my_guide_rate 4058.12
 619  name OPL4 total_guide_rate 0 my_guide_rate 6156.69
```

totto82 commented:
check if #3413 fix your issue

GitPaean commented:
> check if OPM/opm-simulators#3413 fix your issue

Will do. Thanks.

GitPaean commented:
It looks like Jenkins will be happy this time. Really not sure what was happening in the previous jenkins running. The commits looked correct to me.

GitPaean commented:
Since Jenkins looks happy, and the testing locally also good. I am addressing the reviewing comments and squashing the commits now to see if it is a proper trade-off way to get it work then.

GitPaean commented on opm/parser/eclipse/EclipseState/Schedule/Group/GuideRate.hpp:
```diff
@@ -128,6 +133,7 @@ struct GRValState {
     std::unordered_map<std::string, GRValPtr> values;
     std::unordered_map<pair, double, pair_hash> injection_group_values;
     std::unordered_map<std::string, RateVector > potentials;
+    bool time_to_update {false};
```
This is a good suggestion, I will use `guide_rates_expired` for now. It is easy to change it later if we want to.

GitPaean commented on src/opm/parser/eclipse/EclipseState/Schedule/Group/GuideRate.cpp:
```diff
@@ -334,4 +322,28 @@ double GuideRate::get_grvalue_result(const GRValState& gr) const
+        if (update_time < last_update_time) {
+            last_update_time = update_time;
+        }
+    }
```
It is a good suggestion, will do.

GitPaean replied:
Sorry, there is some difficulty here. Because there might be situations that all the guide rates have zero values. In the final commit, 6552613b3827badb40e13611c589ba9347af00e4 , we are trying to discard all the zero valued guide rates when coming to update time checking.

GitPaean commented:
Will finish cleaning up and rebasing later. Have to go.

GitPaean commented:
jenkins build this opm-simulators=3406 please

GitPaean commented:
jenkins build this opm-simulators=3406 please

GitPaean commented:
I just marked this PR ready for review now. This is basically the way requires the least code change to achieve the goal. As discussed before, a more profound refactoring will be necessary to make this functionality in a more proper shape, which will include a full rewriting and redesign of the function `updateGuideRates` from the PR https://github.com/OPM/opm-simulators/pull/3406 , which might also trigger quite some refactoring from the opm-common side. Not totally sure yet.

Anyway, please take a look at the current way in this PR and suggest whether this is an acceptable approach for the problem.  In this PR, basically, we are ignoring the time stamp from the zero-valued guide rate and always update them in case the well is opened or begins getting non-zero valued guide rate. 

From https://github.com/OPM/opm-simulators/pull/3406, the guide rate related update functions are put into one single `updateGuideRates` function to make sure the function `updateGuideRateExpiration` only be called once during the update of guide rates.

bska commented on src/opm/parser/eclipse/EclipseState/Schedule/Group/GuideRate.cpp:
```diff
@@ -334,4 +326,27 @@ double GuideRate::get_grvalue_result(const GRValState& gr) const
+        return;
+    }
+
+    // getting the last update time
+    double last_update_time = std::numeric_limits<double>::max();
+    for ([[maybe_unused]] const auto& [wgname, value] : this->values) {
+        const double update_time = value->curr.sim_time;
+        if (value->curr.value > 0 && update_time < last_update_time) {
+            last_update_time = update_time;
+        }
+    }
```
You can still do this with an algorithm:
```C++
auto previous_update_pos = std::min_element(this->values.begin(), this->values.end(),
    [](const auto& v1, const auto& v2)
{
    return (v1.second->curr.value > 0.0)
        && (v1.second->curr.sim_time < v2.second->curr.sim_time);
});

if ((previous_update_pos == this->values.end()) ||
    !(previous_update_pos->second->curr.value > 0.0))
{
    this->guide_rates_expired = false;
    return;
}

const auto previous_global_update_time = previous_update_pos->second->curr.sim_time;
const auto update_delay = config.model().update_delay();
this->guide_rates_expired = ! (sim_time < (previous_global_update_time + update_delay));
```

bska commented:
I'm going to merge this into master now.  I suspect we'll revise this code later.

GitPaean commented:
> I suspect we'll revise this code later.

Yes. I think so.  I hope to get the results straight for the targeting case first, then the refactoring is more relaxing in some way. Getting all the group controlled cases pass sometimes is not a trivial thing because of all the corner situations. 

The refactoring process for this part should try to not changing any running results in any way unless the reference result is wrong by itself.

GitPaean commented:
thanks for merging.
